### PR TITLE
Add declare target subroutine test from CORAL-2 (issue #60)

### PR DESCRIPTION
**Pull request by jhdavis8**
See issue #60. These tests are checking the behavior of the declare target clause, derived from tests on page 116 of the examples document. They are expected to fail of xlf, although I have not checked this as Summit is down today. 

Both involve calling subroutines on the device, which perform some simple math which is then checked. One checks calling a subroutine from a module, and the other a standalone subroutine.

Right now the module test is passing gfortran on fatnode, while the subroutine test won't compile, complaining about the lack of an explicit interface at the `declare target (test_declare)`, but based on the spec and examples document I believe this test is compliant and gfortran is wrong. If I add an interface statement, the test passes, but the examples document clearly states that the `declare target` is supposed to alleviate that requirement.

### CODE REVIEW

File: tests/4.5/application_kernels/declare_target_module.F90 [hunk 1]
```diff
@@ -0,0 +1,52 @@
+!===--- declare_target_module.F90 ------------------------------------------===//
+!
+! OpenMP API Version 4.5 Nov 2015
+!
+! This test uses the declare target construct within a subroutine, called
+! in a target region. It checks the result of a simple offloaded addition.
+! This test was provided by the LLNL FGPU repository, derived from the
+! OpenMP 4.5 examples document.
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/declare_target_module.F90]: Add link to original repo?

File: tests/4.5/application_kernels/declare_target_module.F90 [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  implicit none
+  INTEGER:: x
+
+  OMPVV_TEST_OFFLOADING
+
+  OMPVV_TEST_SHARED_ENVIRONMENT
+
+  x = N
+
+  !$omp target if(x > THRESHOLD) map(tofrom: x)
```
Comment (josemonsalve2) [inline tests/4.5/application_kernels/declare_target_module.F90]: Do we know why the THRESHOLD Thingy?

Comment (jhdavis8) [reply tests/4.5/application_kernels/declare_target_module.F90]: This is from the original 4.5 examples test, https://www.openmp.org/wp-content/uploads/openmp-examples-4.5.0.pdf, pg  115-116 (122-123 on the pdf viewer). It isn't necessary for this test, however.

Comment (josemonsalve2) [reply tests/4.5/application_kernels/declare_target_module.F90]: Let's leave it like this and blame it on the Examples

**Comment by spophale** [comment]
declare_target_module.F90  compiles and runs correctly with xlc on Summit.
declare_target_subroutine.F90 does not compile. See output below:

** ompvv_lib   === End of Compilation 1 ===
"declare_target_subroutine.F90", line 26.24: 1514-480 (S) The entity test_declare that appears in the DECLARE TARGET directive must be declared in the same scoping unit.
** declare_target_subroutine   === End of Compilation 2 ===
** test_declare   === End of Compilation 3 ===
1501-511  Compilation failed for file declare_target_subroutine.F90.

**Comment by jhdavis8** [comment]
Split into two PRs

File: tests/4.5/application_kernels/declare_target_module.F90 [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+#define N 1024
+
+MODULE module_declare
+CONTAINS
+  SUBROUTINE test_declare(x)
+    INTEGER:: x, m
+    !$omp declare target
+    x = x + 1
+  END SUBROUTINE test_declare
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/declare_target_module.F90]: Try putting the subroutine inside the program

File: tests/4.5/application_kernels/declare_target_subroutine.F90 [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+! Thanks to David Richards and Aaron Black for providing this test.
+!
+!===------------------------------------------------------------------------===//
+
+
+#include "ompvv.F90"
+
+#define N 1024
+
+PROGRAM declare_target_subroutine
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/declare_target_subroutine.F90]: Add CONTAINS to move SUBROUTINE inside the program scope.

Comment (jhdavis8) [reply tests/4.5/application_kernels/declare_target_subroutine.F90]: Tried this, no change to results or the error message.

Comment (jhdavis8) [reply tests/4.5/application_kernels/declare_target_subroutine.F90]: Need to ping the authors about this test to see if anything has changed

Comment (jhdavis8) [reply tests/4.5/application_kernels/declare_target_subroutine.F90]: Also check with the examples document. Consider sending a note to vendors.

Comment (tob2) [reply tests/4.5/application_kernels/declare_target_subroutine.F90:18]: "test_declare" needs an explicit interface (or be in a use-assocated module or being a contained procedure) as it uses a deferred-shape array argument.

If you don't want to make it a contained procedure or put into a module, you could use something like:
```Fortran
INTERFACE
  SUBROUTINE test_declare(a, sum)
    !$omp declare target(test_declare)
    INTEGER,POINTER,INTENT(in) :: a(:)
    INTEGER,INTENT(inout):: sum
  END SUBROUTINE
END INTERFACE
```

Comment (jhdavis8) [reply tests/4.5/application_kernels/declare_target_subroutine.F90:23]: Thanks @tob2! This resolves it.

**Comment by jhdavis8** [comment]
Test passes gfortran and xlf on summit.

**Review by spophale — APPROVED** [review]

**Pull request merged**